### PR TITLE
os/bluestore: cache relevant cleanups

### PR DESCRIPTION
**Pull request by xiexingguo**

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -663,12 +665,12 @@ void BlueStore::BufferSpace::read(
   res.clear();
   uint64_t end = offset + length;
   for (auto i = _data_lower_bound(offset);
-       i != buffer_map.end() && offset < end && i->first < end;
+       i != buffer_map.end() && offset < end && i->first < end && length;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: No need for that addition - the length == 0 case is properly handled by offset < end

**Comment by xiexingguo** [comment]
@liewegas Is the following failure relevant?

```
2016-06-20 12:36:38.231789 2ad67e1f5100 -1 load failed dlopen(/home/jenkins-build/build/workspace/ceph-pull-requests/src/.libs/compressor/libceph_snappy.so): /home/jenkins-build/build/workspace/ceph-pull-requests/src/.libs/compressor/libceph_snappy.so: cannot open shared object file: No such file or directory
2016-06-20 12:36:38.231802 2ad67e1f5100 -1 create cannot load compressor of type snappy
2016-06-20 12:36:38.231804 2ad67e1f5100 -1 bluestore(testdir/test-ceph-disk/dir) _set_compression unable to initialize snappy compressor
 (retry)
command: Running command: /usr/bin/timeout 300 ceph-osd --cluster ceph --mkfs --mkkey -i 0 --monmap testdir/test-ceph-disk/dir/activate.monmap --osd-data testdir/test-ceph-disk/dir --osd-uuid cb2dc548-7886-467a-8d73-2165d6c42c94 --keyring testdir/test-ceph-disk/dir/keyring --setuser root --setgroup root
tests/ceph-disk.sh:302: test_activate_dir_bluestore:  return 1
```

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -670,7 +670,7 @@ void BlueStore::BufferSpace::read(
     Buffer *b = i->second.get();
     assert(b->end() > offset);
     if (b->is_writing() || b->is_clean()) {
-      if (b->offset < offset) {
+      if (b->offset <= offset) {
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: this was deliberate.. the final if/else below handles the b->offset == offset case

**Comment by liewegas** [comment]
@xiexingguo not relevant. can you drop that one boundary condition patch?  then i'll merge.  thanks!

**Comment by xiexingguo** [comment]
@liewegas As you wish, thanks for the confirmation!

File: src/os/bluestore/BlueStore.h [hunk 3]
```diff
@@ -244,6 +244,7 @@ class BlueStore : public ObjectStore,
     }
 
     bool empty() const {
+      std::lock_guard<std::mutex> l(cache->lock);
       return buffer_map.empty();
     }
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.h]: This isn't right--we deadlock with this patch.  (Please run ./ceph_test_objectstore --gtest_filter=*/2 for any bluestore changes!).  The cache->lock protects LRU state, not the forward mapping structures in BufferSpace and OnodeSpace.  Please drop this patch!

**Comment by xiexingguo** [comment]
Done, dropped that patch too. Sorry for the inconvenience.

**Comment by liewegas** [comment]
@xiexingguo are you on irc.oftc.net in #ceph-devel ?

**Comment by xiexingguo** [comment]
Not yet. Is there anything I might be helpful to? And if there is, I'll try to get ready for that too.

**Pull request merged**